presto-hive/src/main/java/com/facebook/presto/hive/HiveMaterializedViewUtils.java

```diff
@@ -0,0 +1,269 @@
+                throw new PrestoException(HIVE_INVALID_METADATA, String.format("Expected %d partition key values, but got %d", partitionColumns.size(), partitionSpec.size()));
+            }
+            for (Column partitionColumn : partitionColumns) {
+                String name = partitionColumn.getName();
+                HiveType hiveType = partitionColumn.getType();
+                if (!hiveType.isSupportedType()) {
+                    throw new PrestoException(NOT_SUPPORTED, String.format("Unsupported Hive type %s found in partition keys of table %s.%s", hiveType, table.getDatabaseName(), table.getTableName()));
+                }
+                if (partitionSpec.get(name) == null) {
+                    throw new PrestoException(HIVE_INVALID_PARTITION_VALUE, String.format("partition key value cannot be null for field: %s", name));
+                }
+            }
+
+            Map<String, NullableValue> partitionNameAndValuesMap = partitionSpec.entrySet().stream().collect(toMap(
+                    Map.Entry::getKey,
+                    entry -> getPartitionValue(entry.getValue())));
+
+            TupleDomain<String> tupleDomain = TupleDomain.fromFixedValues(partitionNameAndValuesMap);
+            partitionNamesAndValues.add(tupleDomain);
+        }
+
+        List<String> partitionColumnNames = partitionColumns.stream().map(Column::getName).collect(toList());
+        return new MaterializedDataPredicates(partitionNamesAndValues, partitionColumnNames);
+    }
+
+    private static NullableValue getPartitionValue(String value)
+    {
+        if (HIVE_DEFAULT_DYNAMIC_PARTITION.equals(value)) {
+            return NullableValue.asNull(VARCHAR);
+        }
+        return NullableValue.of(VARCHAR, utf8Slice(value));
+    }
+
+    public static MaterializedDataPredicates differencePredicatesInfo(
+            MaterializedDataPredicates leftPredicatesInfo,
+            MaterializedDataPredicates rightPredicatesInfo,
+            Map<String, String> rightToLeftPredicatesKeyMap)
+    {
+        if (rightToLeftPredicatesKeyMap.isEmpty()) {
+            return emptyMaterializedViewPredicatesInfo();
+        }
+        if (rightPredicatesInfo.isEmpty()) {
+            return leftPredicatesInfo;
+        }
+
+        Set<String> leftMappedCommonKeys = new HashSet<>();
+        Set<String> rightMappedCommonKeys = new HashSet<>();
+        for (String rightKey : rightPredicatesInfo.getColumnNames()) {
+            String leftKey = rightToLeftPredicatesKeyMap.get(rightKey);
+            if (leftKey != null && leftPredicatesInfo.getColumnNames().contains(leftKey)) {
+                leftMappedCommonKeys.add(leftKey);
+                rightMappedCommonKeys.add(rightKey);
+            }
+        }
+        if (leftMappedCommonKeys.isEmpty()) {
+            return emptyMaterializedViewPredicatesInfo();
+        }
+
+        Set<Map<String, NullableValue>> rightPredicatesMappedToLeftCommonKeys = new HashSet<>();
```

Comment:
Let's write our own `equal` and `hashcode` methods for `Map` rather than relying on `Set`. It is dangerous to assume each implementation of `Map` will rely on content equality rather than reference equality.

Reply:
Good catch